In [1]:
pip install fredapi

Note: you may need to restart the kernel to use updated packages.


In [2]:
import pandas as pd

In [3]:
from fredapi import Fred

In [4]:
import json

In [5]:
import requests

In [162]:
import datetime

# 1.1. Extracting Information from my own transaction

In [155]:
answer2 = requests.get("https://api.whatsonchain.com/v1/bsv/main/block/height/668416")
bitcointransfer_block = json.loads(answer2.text)


In [156]:
bitcointransfer_block

{'hash': '000000000000000005852a48b413a006086edb0ffc1c80a0cf2e71fb1146636e',
 'confirmations': 23,
 'size': 218998,
 'height': 668416,
 'version': 536870912,
 'versionHex': '20000000',
 'merkleroot': 'e1e100752286e5fa7a50a1b3c6c56bd904f5f4b1bd4bf917bd25469fb0ca18d9',
 'txcount': 414,
 'tx': ['b56ac266b2ce55b2d7e6d6011556e79949cc0059f24633dbe15f096a59892d63',
  '786e8ce2778bfd53f3a66bca4e2c4ec68ac75148334d72bb19813e250902ce1b',
  '16cd061cdc793dacad62ecb0ae25e443e2f93c1cbfaf3d7c0c0c0a9b5212d166',
  '93cd0b3e2f878dfabc83f9d08f4fba4e0b492ae8c56ec2dc1adeec4f9e215e45',
  '6e3171247e8d99316c1fd811ea5994c85cdfd7bec2501073f50c1653010fb3fc',
  '6074f8924e4274cd2d209f071248e0492df2df36b91950fa091f82093ed061b4',
  'd5de86d477305546cf8acc683d1e30133dc2f932c86747bcfbfdc02a3ee34725',
  'ba3a4bd95c123f8ace005535434ca3a2c7d0c23f662ed8fb4b8d2e7481ef251e',
  '6428225ecb0d7530de9d8d12a479ce2bc41a44e3d2973885ed8db10895814cae',
  'c4d7d87afddc2c591d09b483ee887f8e48a96ff7bd925603546a8071b13c76d2',
  '183ca0

In [157]:
print(bitcointransfer_block['miner'])

ViaBTC


In [ ]:
# Miners can be defined as accountants who records every transactions to the blockchain. The concept is a proof of payment, which is important if we want your payment to be valid. Here the miner name is ViaBTC

In [158]:
print(bitcointransfer_block['txcount'])

414


In [ ]:
# Total number of transaction done since this transaction has taken place.

In [159]:
print(bitcointransfer_block['totalFees'])

0.0014973700000000534


In [ ]:
#total transaction that has been carried out including the mining fee in bitcoin values

In [160]:
print(bitcointransfer_block['confirmations'])

23


In [ ]:
# The concept of "confirmation" in Bitcoin essentially translates to how many blocks have been generated since the block containing transaction has been created.

In [164]:
timestamp = datetime.datetime.fromtimestamp(bitcointransfer_block['time'])

In [165]:
print(timestamp.strftime('%Y-%m-%d %H:%M:%S'))

2021-01-03 18:38:44


In [ ]:
# timestamp essentially tells us the time of the bitcoin transaction. Here itwas done on 3rd January 2021 at 6:38 p.m.

# 1.2  Extract Information from Famous Blocks

The First ever transaction from Satoshi to Hal Finney in 2010

In [6]:
answer = requests.get("https://api.whatsonchain.com/v1/bsv/main/block/height/170")
bitcoin_block = json.loads(answer.text)


In [7]:
bitcoin_block

{'hash': '00000000d1145790a8694403d4063f323d499e655c83426834d4ce2f8dd4a2ee',
 'confirmations': 668195,
 'size': 490,
 'height': 170,
 'version': 1,
 'versionHex': '00000001',
 'merkleroot': '7dac2c5666815c17a3b36427de37bb9d2e2c5ccec3f8633eb91a4205cb4c10ff',
 'txcount': 2,
 'tx': ['b1fea52486ce0c62bb442b530a3f0132b826c74e473d1f2c220bfa78111c5082',
  'f4184fc596403b9d638783cf57adfe4c75c605f6356fbc91338530e9831e9e16'],
 'time': 1231731025,
 'mediantime': 1231716245,
 'nonce': 1889418792,
 'bits': '1d00ffff',
 'difficulty': 1,
 'chainwork': '000000000000000000000000000000000000000000000000000000ab00ab00ab',
 'previousblockhash': '000000002a22cfee1f2c846adbd12b3e183d4f97683f85dad08a79780a84bd55',
 'nextblockhash': '00000000c9ec538cab7f38ef9c67a95742f56ab07b0a37c5be6b02808dbfb4e0',
 'coinbaseTx': {'hex': '',
  'txid': 'b1fea52486ce0c62bb442b530a3f0132b826c74e473d1f2c220bfa78111c5082',
  'hash': 'b1fea52486ce0c62bb442b530a3f0132b826c74e473d1f2c220bfa78111c5082',
  'version': 1,
  'size': 134,

In [8]:
print(bitcoin_block['miner'])

��


In [166]:
print(bitcoin_block['txcount'])

2


In [167]:
print(bitcoin_block['totalFees'])

0


In [168]:
print(bitcoin_block['confirmations'])

668195


In [169]:
timestamp1 = datetime.datetime.fromtimestamp(bitcoin_block['time'])

In [170]:
print(timestamp1.strftime('%Y-%m-%d %H:%M:%S'))

2009-01-12 03:30:25


The Pizza purchase for 10,000BTC in 2010

In [11]:
answer1 = requests.get("https://api.whatsonchain.com/v1/bsv/main/block/height/57043")
pizza_block = json.loads(answer1.text)

In [12]:
pizza_block

{'hash': '00000000152340ca42227603908689183edc47355204e7aca59383b0aaac1fd8',
 'confirmations': 611322,
 'size': 23835,
 'height': 57043,
 'version': 1,
 'versionHex': '00000001',
 'merkleroot': '5c1d2211f598cd6498f42b269fe3ce4a6fdb40eaa638f86a0579c4e63a721b5a',
 'txcount': 2,
 'tx': ['bd9075d78e65a98fb054cb33cf0ecf14e3e7f8b3150231df8680919a79ac8fe5',
  'a1075db55d416d3ca199f55b6084e2115b9345e16c5cf302fc80e9d5fbf5d48d'],
 'time': 1274552191,
 'mediantime': 1274548928,
 'nonce': 188133155,
 'bits': '1c159c24',
 'difficulty': 11.84622814970471,
 'chainwork': '00000000000000000000000000000000000000000000000000026c60b6429108',
 'previousblockhash': '0000000013e7e85518dac94d012d73253d3fdac5c30c4143b177f3086f129580',
 'nextblockhash': '0000000013ab9f8ed78b254a429d3d5ad52905362e01bf6c682940337721eb51',
 'coinbaseTx': {'hex': '',
  'txid': 'bd9075d78e65a98fb054cb33cf0ecf14e3e7f8b3150231df8680919a79ac8fe5',
  'hash': 'bd9075d78e65a98fb054cb33cf0ecf14e3e7f8b3150231df8680919a79ac8fe5',
  'version'

In [13]:
print(pizza_block['miner'])

$�v


In [171]:
print(pizza_block['txcount'])

2


In [172]:
print(pizza_block['totalFees'])

0.990000000000002


In [173]:
print(pizza_block['confirmations'])

611322


In [174]:
timestamp2 = datetime.datetime.fromtimestamp(pizza_block['time'])

In [175]:
print(timestamp2.strftime('%Y-%m-%d %H:%M:%S'))

2010-05-22 19:16:31


# 2. Time Series Investigation of Bitcoin Price

In [16]:
CBBTCUSD1 = fred1.get_series('CBBTCUSD')


In [17]:
bitcoinprice = CBBTCUSD1

In [71]:
goldprice = fred1.get_series('ID7108')
market = fred1.get_series('SP500')
riskfreerate = fred1.get_series('TB3MS')

In [72]:
dataframe = pd.concat([bitcoinprice, goldprice, market, riskfreerate], axis=1)

In [73]:
dataframe.columns = ['bitcoinprice', 'goldprice', 'market', 'riskfreerate']

In [74]:
dataframe

bitcoinprice  goldprice   market  riskfreerate
1934-01-01           NaN        NaN      NaN          0.72
1934-02-01           NaN        NaN      NaN          0.62
1934-03-01           NaN        NaN      NaN          0.24
1934-04-01           NaN        NaN      NaN          0.15
1934-05-01           NaN        NaN      NaN          0.16
...                  ...        ...      ...           ...
2020-12-29      27407.85        NaN  3727.04           NaN
2020-12-30      28934.56        NaN  3732.04           NaN
2020-12-31      29026.97        NaN  3756.07           NaN
2021-01-01      29391.09        NaN      NaN           NaN
2021-01-02      32244.73        NaN      NaN           NaN

[4184 rows x 4 columns]

In [79]:
dataframe.to_csv('btc_analysis.csv')

In [76]:
dataframe = dataframe.fillna(method='ffill')

In [78]:
dataframe = dataframe.loc['2016-01-01':'2021-01-01']

In [80]:
dataframe

bitcoinprice  goldprice   market  riskfreerate
2016-01-01        435.66      395.0  2043.94          0.26
2016-01-02        435.40      395.0  2043.94          0.26
2016-01-03        431.91      395.0  2043.94          0.26
2016-01-04        433.85      395.0  2012.66          0.26
2016-01-05        433.34      395.0  2016.71          0.26
...                  ...        ...      ...           ...
2020-12-28      27044.14      694.6  3735.36          0.09
2020-12-29      27407.85      694.6  3727.04          0.09
2020-12-30      28934.56      694.6  3732.04          0.09
2020-12-31      29026.97      694.6  3756.07          0.09
2021-01-01      29391.09      694.6  3756.07          0.09

[1828 rows x 4 columns]

In [81]:
dataframe = dataframe[dataframe.index.day==1]

In [82]:
dataframe

bitcoinprice  goldprice   market  riskfreerate
2016-01-01        435.66      395.0  2043.94          0.26
2016-02-01        371.33      412.4  1939.38          0.31
2016-03-01        433.08      451.2  1978.35          0.29
2016-04-01        417.68      445.4  2072.78          0.23
2016-05-01        456.98      465.7  2065.30          0.27
...                  ...        ...      ...           ...
2020-09-01      11911.74      707.7  3526.65          0.11
2020-10-01      10598.64      690.2  3380.80          0.10
2020-11-01      13784.52      694.6  3269.96          0.09
2020-12-01      18763.35      694.6  3662.45          0.09
2021-01-01      29391.09      694.6  3756.07          0.09

[61 rows x 4 columns]

In [102]:
import numpy as np

In [84]:
dataframe['Rbt'] = np.log(dataframe['bitcoinprice']/dataframe['bitcoinprice'].shift(1))

<ipython-input-84-5d5e015119e5>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['Rbt'] = np.log(dataframe['bitcoinprice']/dataframe['bitcoinprice'].shift(1))


In [86]:
dataframe['Rgt'] = np.log(dataframe['goldprice']/dataframe['goldprice'].shift(1))
dataframe['Rmt'] = np.log(dataframe['market']/dataframe['market'].shift(1))
dataframe['Rft'] = dataframe['riskfreerate']/4

<ipython-input-86-e92e98fb25bc>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['Rgt'] = np.log(dataframe['goldprice']/dataframe['goldprice'].shift(1))
<ipython-input-86-e92e98fb25bc>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dataframe['Rmt'] = np.log(dataframe['market']/dataframe['market'].shift(1))
<ipython-input-86-e92e98fb25bc>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the 

In [87]:
dataframe

bitcoinprice  goldprice   market  riskfreerate       Rbt  \
2016-01-01        435.66      395.0  2043.94          0.26       NaN   
2016-02-01        371.33      412.4  1939.38          0.31 -0.159771   
2016-03-01        433.08      451.2  1978.35          0.29  0.153831   
2016-04-01        417.68      445.4  2072.78          0.23 -0.036207   
2016-05-01        456.98      465.7  2065.30          0.27  0.089924   
...                  ...        ...      ...           ...       ...   
2020-09-01      11911.74      707.7  3526.65          0.11  0.008932   
2020-10-01      10598.64      690.2  3380.80          0.10 -0.116799   
2020-11-01      13784.52      694.6  3269.96          0.09  0.262821   
2020-12-01      18763.35      694.6  3662.45          0.09  0.308359   
2021-01-01      29391.09      694.6  3756.07          0.09  0.448786   

                 Rgt       Rmt     Rft  
2016-01-01       NaN       NaN  0.0650  
2016-02-01  0.043108 -0.052511  0.0775  
2016-03-01  0.089917  0.019895  0.0725  
2016-04-01 -0.012938  0.046628  0.0575  
2016-05-01  0.044569 -0.003615  0.0675  
...              ...       ...     ...  
2020-09-01 -0.032120  0.075216  0.0275  
2020-10-01 -0.025039 -0.042236  0.0250  
2020-11-01  0.006355 -0.033335  0.0225  
2020-12-01  0.000000  0.113355  0.0225  
2021-01-01  0.000000  0.025241  0.0225  

[61 rows x 8 columns]

In [128]:
dataframe = dataframe.drop(dataframe.index[0])

In [129]:
Bitcoinreturns = dataframe["Rbt"]    
Goldreturns = dataframe["Rgt"]
Marketreturns = dataframe["Rmt"]
Riskfreereturns = dataframe["Rft"]

In [130]:
Bit_Gold_correl = Bitcoinreturns.corr(Goldreturns)

In [131]:
Bit_Gold_correl

0.23583022187096742

The correlation coefficient between bitcoinreturn and goldreturn is 0.236, which says that both the returns are positively weakly related with each other.

In [132]:
Bit_Market_correl = Bitcoinreturns.corr(Marketreturns)

In [133]:
Bit_Market_correl

0.19089529681473805

Bitcoinreturns and Marketreturns are very weakly positively correlated with each other with a value of 0.1908.

In [134]:
!pip3 install scikit-learn

In [135]:
from sklearn.linear_model import LinearRegression

In [136]:
w = Bitcoinreturns-Riskfreereturns
x = Goldreturns-Riskfreereturns
y = Marketreturns-Riskfreereturns

In [139]:
w.shape

(59,)

In [142]:
w = w.values.reshape(-1, 1)

In [143]:
w.shape

(59, 1)

In [148]:
w

array([[ 0.08133131],
       [-0.09370688],
       [ 0.02242404],
       [ 0.08982801],
       [ 0.16245985],
       [-0.1888685 ],
       [-0.12713836],
       [-0.010742  ],
       [ 0.05882436],
       [-0.09611061],
       [ 0.14778794],
       [-0.13020144],
       [ 0.02929058],
       [-0.31900329],
       [ 0.05169872],
       [ 0.27655381],
       [-0.26599699],
       [-0.13251654],
       [ 0.33623156],
       [-0.38605713],
       [ 0.1265969 ],
       [ 0.14147078],
       [-0.13728413],
       [-0.79490341],
       [-0.23541603],
       [-0.90911415],
       [-0.17904954],
       [-0.66341418],
       [-0.65843446],
       [-0.32720296],
       [-0.58833174],
       [-0.65239688],
       [-0.617935  ],
       [-1.01882831],
       [-0.67183517],
       [-0.70460154],
       [-0.49812103],
       [-0.51165255],
       [-0.33578144],
       [-0.06732176],
       [-0.31378923],
       [-0.50211633],
       [-0.53668251],
       [-0.57270519],
       [-0.27750682],
       [-0

In [144]:
x = x.values.reshape(-1, 1)

In [145]:
y = y.values.reshape(-1, 1)

In [151]:
lr = LinearRegression()

In [153]:
lr.fit(w, y)
print(lr.intercept_, lr.coef_)

[-0.17190718] [[0.4791191]]


The intercept term for Bitcoinreturns is -0.1719 and the slope coeeficient is 0.4791

In [154]:
lr.fit(x, y)
print(lr.intercept_, lr.coef_)

[-0.00441081] [[0.97531276]]


The intercept term for goldreturn is -0.0044 and the slope coeeficent for goldreturns is 0.9753